runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -1065,12 +1072,14 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 	fixupDataAddr(MM_ForwardedHeader *forwardedHeader, omrobjectptr_t arrayPtr)
 	{
 #if defined(J9VM_ENV_DATA64)
-		J9IndexableObject *j9ArrayPtr = (J9IndexableObject *)arrayPtr;
+		if (static_cast<J9JavaVM*>(_omrVM->_language_vm)->isDataAddrPresent) {
```

Comment:
This is good for this pull request to keep things simple, but eventually for perf benefits we should try to remove this check from here and shuffle the code somehow (through code cloning/specialization) so that Scavenger does not end up calling fixupDataAddr, to start with.

Reply:
Seems like a specialization has already been done for other purposes, namely Balanced, unlike Scavenger, needs to fixup leaf pointers. So, Balanced actually does not use generic GC_ObjectModel::fixupForwardedObject(), but has its own clone of that functionality. 

That means, as far as Scavenger-vs-CopyForward is concerned we should be able to remove _indexableObjectModel->fixupDataAddr() from GC_ObjectModel::fixupForwardedObject(), and then remove isDataAddrPresent check from fixupDataAddr().

However, indexableObjectModel.fixupDataAddr() is used for sliding compactors too, and both Standard GC and Balanced use the same API: MM_EnvironmentDelegate::postObjectMoveForCompact(). Balanced still needs to do extra fixup of leaf pointers there, and the specialization is done by checking isVLHGC(). What we need to there is just to pull isVLHGC() check in front of isIndexable() check, so we do neither dataAddr nor leaf pointer fixup, for Standard GCs. Since sliding compactor is rarely used, it's fine to keep isVLHGC check there (for the sake of having fixupDataAddr free of isDataAddrPresent).

While possibly we could do these changes even as a part of this pull request, it's cleaner/simpler to follow it up as a separate PR.

Reply:
Ok 👍